firmware/library/L4_Application/task_scheduler.cpp

```diff
@@ -0,0 +1,109 @@
+#include "L4_Application/task_scheduler.hpp"
+
+#include <cstring>
+
+#include "L4_Application/task.hpp"
+
+void rtos::TaskScheduler::RunTask(void * task_ptr)
+{
+    TaskInterface & task = *(reinterpret_cast<TaskInterface *>(task_ptr));
+
+    const uint8_t kTaskCount = TaskScheduler::Instance().GetTaskCount();
+    const uint8_t kTaskIndex =
+        TaskScheduler::Instance().GetTaskIndex(task.GetName());
+    SJ2_ASSERT_FATAL(kTaskIndex < kTaskCount,
+                     "The task index should not exceed the task count.");
+
+    EventGroupHandle_t pre_run_event_group_handle =
+        TaskScheduler::Instance().GetPreRunEventGroupHandle();
+    const EventBits_t kPreRunSyncBits =
+        TaskScheduler::Instance().GetPreRunSyncBits();
+    const uint32_t kSyncBit = (1 << kTaskIndex);
+    SJ2_ASSERT_FATAL(task.PreRun(),
+                     "PreRun() failed for task: %s, terminating scheduler!",
+                     task.GetName());
+    // wait for all other PreRun() of other tasks to finish
+    xEventGroupSync(pre_run_event_group_handle, kSyncBit, kPreRunSyncBits,
+                    portMAX_DELAY);
+    // ALl PreRun() complete, each Task's Run() can now start executing...
+    TickType_t last_wake_time = xTaskGetTickCount();
+    while (1)
+    {
+        if (!task.Run())
+        {
+            SJ2_ASSERT_WARNING(
+                false,
+                "An error occured, the following task will be suspended: %s",
+                task.GetName());
+            vTaskSuspend(NULL);
+        }
+        // delay task if the task's delay time has been set...
+        uint32_t delay_time = task.GetDelayTime();
+        if (delay_time)
+        {
+            vTaskDelayUntil(&last_wake_time, delay_time);
+        }
+    }
+};
+void rtos::TaskScheduler::Start()
+{
+    SJ2_ASSERT_FATAL(
+        task_count_ != 0,
+        "Attempting to start TaskScheduler but there are no tasks that "
+        "are currently scheduled.");
+    InitializeAllTasks();
+    vTaskStartScheduler();
+};
+
+void rtos::TaskScheduler::InitializeAllTasks()
+{
+    for (uint32_t i = 0; i < task_count_; i++)
+    {
+        TaskInterface * task = task_list_[i];
+        *(task->GetHandle()) = xTaskCreateStatic(
+            RunTask, task->GetName(),
```

Comment:
Nit: I think that having each parameter in its own line would look cleaner, even if `clang-format` disagrees.